pkg/apis/networking/v1alpha1/domain_types.go

```diff
@@ -0,0 +1,160 @@
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Domain is a cluster-scoped resource to configure a proxy pool for a given Route.
+type Domain struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the desired state of the Domain.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec DomainSpec `json:"spec,omitempty"`
+
+	// Status is the current state of the Domain.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status DomainStatus `json:"status,omitempty"`
+}
+
+// Verify that Domain adheres to the appropriate interfaces.
+var (
+	// Check that Doimain may be validated and defaulted.
+	_ apis.Validatable = (*Domain)(nil)
+	_ apis.Defaultable = (*Domain)(nil)
+
+	// Check that we can create OwnerReferences to a Domain.
+	_ kmeta.OwnerRefable = (*Domain)(nil)
+
+	// Check that the type conforms to the duck Knative Resource shape.
+	_ duckv1.KRShaped = (*Domain)(nil)
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// DomainList is a collection of Domain objects.
+type DomainList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// Items is the list of Domain objects.
+	Items []Domain `json:"items"`
+}
+
+// DomainSpec describes the Ingress the user wishes to exist.
+//
+// In general this follows the same shape as K8s Ingress.
+// Some notable differences:
+// - Backends now can have namespace:
+// - Traffic can be split across multiple backends.
+// - Timeout & Retry can be configured.
+// - Headers can be appended.
+// DomainSpec contains the specification of the Domain CRD.
+type DomainSpec struct {
+	// IngressClass tells what Ingress class annotation to use for Routes associated
+	// with this Realm.
+	IngressClass string `json:"ingressClass"`
+
+	// Suffix specifies the domain suffix to be used. This field replaces the
+	// existing config-domain ConfigMap.  Internal Domains can omit this, in
+	// which case we will default to the cluster suffix.
+	// +optional
+	Suffix string `json:"suffix,omitempty"`
+
+	// LoadBalancers provide addresses (IP address, domains) of the load balancers
+	// associated with this Domain.  This is used in automatic DNS provisioning like
+	// configuration of magic DNS or creating ExternalName services for cluster-local
+	// access.
+	LoadBalancers []LoadBalancerIngressSpec `json:"loadBalancers"`
+
+	// Config represents multiple resource information required for Kingress implementation.
+	// +optional
+	Configs []IngressConfig `json:"configs,omitempty"`
```

Comment:
the comment says `required` but it looks it's optional?

Reply:
It is an optional field if the specific implementation chooses not to use this field. In certain implementations like Istio this is a required field. We will remove `required` to avoid the confusion